### PR TITLE
[RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57288 [RPC Framework] Enable RemoteModule to directly send GPU tensors over the wire on TensorPipe RPC backend if a device map is provided**

If the device map provided by RemoteModue is not empty, then TensorPipe RPC backend can support directly sending GPU tensors over the wire.

Also add pybind of `_get_device_map`.

The changes in unit test setup is separated out as a follow-up PR, as currently it breaks some tests in distributed/rpc/test_faulty_agent.py.

#Original PR issue: https://github.com/pytorch/pytorch/issues/51670

Differential Revision: [D28021672](https://our.internmc.facebook.com/intern/diff/D28021672/)

### CODE REVIEW

File: torch/distributed/nn/jit/templates/remote_module_template.py [hunk 1]
```diff
@@ -17,47 +17,46 @@
 # WARNING: If the module is on a cuda device, any CPU tensor stored in a nested ``arg`` or `kwargs``,
 # will not be implicitly moved to the same cuda device.
 def _remote_forward(
-    module_rref: RRef[module_interface_cls], device: str, {arg_types}){arrow_and_return_type}:
+    module_rref: RRef[module_interface_cls], device: str, is_device_map_set: bool, {arg_types}){arrow_and_return_type}:
```
Comment (mrshenli) [inline torch/distributed/nn/jit/templates/remote_module_template.py:20]: Is `is_device_map_set` argument user-facing?

Comment (SciPioneer) [reply torch/distributed/nn/jit/templates/remote_module_template.py:20]: Not user-facing. See L70. We pass a wrapper of (internal args, user-facing args) into `_remote_forward`.

**Review by wanchaol — APPROVED** [review]

**Review by rohan-varma — COMMENTED** [review]
(Can be done in a follow up PR)
Can the docs be enhanced to mention that inputs are automatically moved to the right device on the remote worker, and , for TensorPipe backend, GPU tensors are returned over the wire according to the device_maps set?

**Comment by SciPioneer** [comment]
> inputs are automatically moved to the right device on the remote worker, and , for TensorPipe backend, GPU tensors are returned over the wire according to the device_maps set?

Thanks! Added some descriptions to RemoteModule class.

**Pull request closed without merging**